src/engine/renderer/tr_shader.cpp

```diff
@@ -5992,209 +5990,188 @@ A second parameter will cause it to print in sorted order
+		iter !=shaderSortStrings.end(); iter++ )
+	{
+		maxShaderSortStringLen = std::max( maxShaderSortStringLen, (int)iter->second.length() );
+	}
+
+	using stageTypeStringMap = std::unordered_map<stageType_t, std::string>;
+	stageTypeStringMap stageTypeStrings;
+	stageTypeStrings[stageType_t::ST_COLORMAP] = "COLORMAP";
+	stageTypeStrings[stageType_t::ST_GLOWMAP] = "GLOWMAP";
+	stageTypeStrings[stageType_t::ST_DIFFUSEMAP] = "DIFFUSEMAP";
+	stageTypeStrings[stageType_t::ST_NORMALMAP] = "NORMALMAP";
+	stageTypeStrings[stageType_t::ST_HEIGHTMAP] = "HEIGHTMAP";
+	stageTypeStrings[stageType_t::ST_PHYSICALMAP] = "PHYSICALMAP";
+	stageTypeStrings[stageType_t::ST_SPECULARMAP] = "SPECULARMAP";
+	stageTypeStrings[stageType_t::ST_REFLECTIONMAP] = "REFLECTIONMAP";
+	stageTypeStrings[stageType_t::ST_REFRACTIONMAP] = "REFRACTIONMAP";
+	stageTypeStrings[stageType_t::ST_DISPERSIONMAP] = "DISPERSIONMAP";
+	stageTypeStrings[stageType_t::ST_SKYBOXMAP] = "SKYBOXMAP";
+	stageTypeStrings[stageType_t::ST_SCREENMAP] = "SCREENMAP";
+	stageTypeStrings[stageType_t::ST_PORTALMAP] = "PORTALMAP";
+	stageTypeStrings[stageType_t::ST_HEATHAZEMAP] = "HEATHAZEMAP";
+	stageTypeStrings[stageType_t::ST_LIQUIDMAP] = "LIQUIDMAP";
+	stageTypeStrings[stageType_t::ST_LIGHTMAP] = "LIGHTMAP";
+	stageTypeStrings[stageType_t::ST_COLLAPSE_lighting_PBR] = "LIGHTING_PBR";
+	stageTypeStrings[stageType_t::ST_COLLAPSE_lighting_PHONG] = "LIGHTING_PHONG";
+	stageTypeStrings[stageType_t::ST_COLLAPSE_reflection_CB] = "REFLECTION_CB";
+	stageTypeStrings[stageType_t::ST_ATTENUATIONMAP_XY] = "ATTENUATIONMAP_XY";
+	stageTypeStrings[stageType_t::ST_ATTENUATIONMAP_Z] = "ATTENUATIONMAP_XZ";
+
+	std::string stageTypeHeader = "stageType";
+	int maxStageTypeStringLen = stageTypeHeader.length();
+
+	for ( stageTypeStringMap::iterator iter = stageTypeStrings.begin();
+		iter !=stageTypeStrings.end(); iter++ )
+	{
+		maxStageTypeStringLen = std::max( maxStageTypeStringLen, (int)iter->second.length() );
+	}
+
+	std::string interactLightHeader = "interactLight";
+	int maxInteractLightStringLen = interactLightHeader.length();
+
+	std::string separator = " ";
+
+	std::string header = Str::Format(
+		"%-*s" "%s%-*s" "%s%-*s" "%s%-*s" "%s%s",
+		maxShaderTypeStringLen, shaderTypeHeader,
+		separator, maxShaderSortStringLen, shaderSortHeader,
+		separator, maxInteractLightStringLen, interactLightHeader,
+		separator, maxStageTypeStringLen, stageTypeHeader,
+		separator, "stageNumber:shaderName" );
+
+	std::string lineSeparator( header.length(), '-' );
+
+	Log::CommandInteractionMessage( lineSeparator );
+	Log::CommandInteractionMessage( header );
+	Log::CommandInteractionMessage( lineSeparator );
+
+	int stageCount = 0;
+	int highestStage = 0;
+
+	for ( int i = 0; i < tr.numShaders; i++ )
+	{
+		shader_t *shader = ri.Cmd_Argc() > 2 ? tr.sortedShaders[ i ] : tr.shaders[ i ];
 
 		if ( s && !Com_Filter( s, shader->name, false ) )
 		{
 			continue;
 		}
 
-		std::string str;
+		std::string foundShaderTypeString = "";
```

Comment:
The point of maps is to efficiently look up the value for a key. `foundShaderTypeString = shaderTypeStrings[shader->type]` (it will automatically be an empty string for missing ones)

Reply:
Oh lol, thanks. I was looking for that (that's why I started using a map) but I searched for ways to get a value with a map and every c++ page I have read was iterating… 🤦‍♀️️ I was like “this can't be true 🤷‍♀️️”! It wasn't indeed.

Reply:
Done.